### PR TITLE
[MM-18676] Add tag annotations to plugin server API methods

### PR DESCRIPTION
**Pull request by cinlloc**
#### Summary
After merging https://github.com/mattermost/mattermost-developer-documentation/pull/383, it will allow to group methods by tag in documentation index.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-18676
Fixes https://github.com/mattermost/mattermost-server/issues/12265

#### Comments
As I haven't a deep knowledge of API methods behavior, this categorization is based on naming+comments. Don't hesitate if you see improvements!

### CODE REVIEW

File: plugin/api.go [hunk 1]
```diff
@@ -40,139 +43,167 @@ type API interface {
...
 	// UpdateUserActive deactivates or reactivates an user.
 	//
+	// @tag User
 	// Minimum server version: 5.8
 	UpdateUserActive(userId string, active bool) *model.AppError
 
 	// GetUsersInChannel returns a page of users in a channel. Page counting starts at 0.
 	// The sortBy parameter can be: "username" or "status".
 	//
+	// @tag User
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag Channel`

File: plugin/api.go [hunk 2]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 
 	// GetTeamByName gets a team by its name.
 	//
+	// @tag Team
 	// Minimum server version: 5.2
 	GetTeamByName(name string) (*model.Team, *model.AppError)
 
 	// GetTeamsUnreadForUser gets the unread message and mention counts for each team to which the given user belongs.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 3]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// GetTeamsForUser returns list of teams of given user ID.
 	//
+	// @tag Team
+	// @tag User
 	// Minimum server version: 5.6
 	GetTeamsForUser(userId string) ([]*model.Team, *model.AppError)
 
 	// CreateTeamMember creates a team membership.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 4]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// Minimum server version: 5.2
 	CreateTeamMember(teamId, userId string) (*model.TeamMember, *model.AppError)
 
 	// CreateTeamMember creates a team membership for all provided user ids.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 5]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// Minimum server version: 5.2
 	CreateTeamMembers(teamId string, userIds []string, requestorId string) ([]*model.TeamMember, *model.AppError)
 
 	// DeleteTeamMember deletes a team membership.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 6]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// Minimum server version: 5.2
 	DeleteTeamMember(teamId, userId, requestorId string) *model.AppError
 
 	// GetTeamMembers returns the memberships of a specific team.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 7]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// Minimum server version: 5.2
 	GetTeamMembers(teamId string, page, perPage int) ([]*model.TeamMember, *model.AppError)
 
 	// GetTeamMember returns a specific membership.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 8]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// GetTeamMembersForUser returns all team memberships for a user.
 	//
+	// @tag Team
+	// @tag User
 	// Minimum server version: 5.10
 	GetTeamMembersForUser(userId string, page int, perPage int) ([]*model.TeamMember, *model.AppError)
 
 	// UpdateTeamMemberRoles updates the role for a team membership.
 	//
+	// @tag Team
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 9]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// GetChannelStats gets statistics for a channel.
 	//
+	// @tag Channel
 	// Minimum server version: 5.6
 	GetChannelStats(channelId string) (*model.ChannelStats, *model.AppError)
 
 	// GetDirectChannel gets a direct message channel.
 	// If the channel does not exist it will create it.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 10]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// Minimum server version: 5.2
 	GetDirectChannel(userId1, userId2 string) (*model.Channel, *model.AppError)
 
 	// GetGroupChannel gets a group message channel.
 	// If the channel does not exist it will create it.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 11]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 
 	// SearchUsers returns a list of users based on some search criteria.
 	//
+	// @tag User
 	// Minimum server version: 5.6
 	SearchUsers(search *model.UserSearch) ([]*model.User, *model.AppError)
 
 	// SearchPostsInTeam returns a list of posts in a specific team that match the given params.
 	//
+	// @tag Post
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag Team`

File: plugin/api.go [hunk 12]
```diff
@@ -181,164 +212,202 @@ type API interface {
...
 	// Minimum server version: 5.10
 	SearchPostsInTeam(teamId string, paramsList []*model.SearchParams) ([]*model.Post, *model.AppError)
 
 	// AddChannelMember joins a user to a channel (as if they joined themselves)
 	// This means the user will not receive notifications for joining the channel.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 13]
```diff
@@ -350,145 +419,175 @@ type API interface {
 
 	// GetChannelMember gets a channel membership for a user.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 14]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// Minimum server version: 5.2
 	GetChannelMember(channelId, userId string) (*model.ChannelMember, *model.AppError)
 
 	// GetChannelMembers gets a channel membership for all users.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 15]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// Minimum server version: 5.6
 	GetChannelMembers(channelId string, page, perPage int) (*model.ChannelMembers, *model.AppError)
 
 	// GetChannelMembersByIds gets a channel membership for a particular User
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 16]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// GetChannelMembersForUser returns all channel memberships on a team for a user.
 	//
+	// @tag Channel
+	// @tag User
 	// Minimum server version: 5.10
 	GetChannelMembersForUser(teamId, userId string, page, perPage int) ([]*model.ChannelMember, *model.AppError)
 
 	// UpdateChannelMemberRoles updates a user's roles for a channel.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 17]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// Minimum server version: 5.2
 	UpdateChannelMemberRoles(channelId, userId, newRoles string) (*model.ChannelMember, *model.AppError)
 
 	// UpdateChannelMemberNotifications updates a user's notification properties for a channel.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 18]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// Minimum server version: 5.2
 	UpdateChannelMemberNotifications(channelId, userId string, notifications map[string]string) (*model.ChannelMember, *model.AppError)
 
 	// DeleteChannelMember deletes a channel membership for a user.
 	//
+	// @tag Channel
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag User`

File: plugin/api.go [hunk 19]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 
 	// GetPost gets a post.
 	//
+	// @tag Post
 	// Minimum server version: 5.2
 	GetPost(postId string) (*model.Post, *model.AppError)
 
 	// GetPostsSince gets posts created after a specified time as Unix time in milliseconds.
 	//
+	// @tag Post
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag Channel`

File: plugin/api.go [hunk 20]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// Minimum server version: 5.6
 	GetPostsSince(channelId string, time int64) (*model.PostList, *model.AppError)
 
 	// GetPostsAfter gets a page of posts that were posted after the post provided.
 	//
+	// @tag Post
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag Channel`

File: plugin/api.go [hunk 21]
```diff
@@ -350,145 +419,175 @@ type API interface {
...
 	// Minimum server version: 5.6
 	GetPostsAfter(channelId, postId string, page, perPage int) (*model.PostList, *model.AppError)
 
 	// GetPostsBefore gets a page of posts that were posted before the post provided.
 	//
+	// @tag Post
```
Comment (iomodo) [inline plugin/api.go]: We could add `@tag Channel`

**Comment by saturninoabril** [comment]
@cinlloc Will look into this once comments above are resolved.

**Comment by cinlloc** [comment]
Thanks. Pushed a new commit with changes requested. Also added some tags to new methods, and a a new "Group" tag. Didn't tag "GetGroupChannel" with "Group" because no Group object involved, but not very sure about it. 
It occurred to me by merging the master that it may be useful to document this tagging feature in some developer documentation part, I will try to find where.

**Comment by hanzei** [comment]
@ali-farooq0 Gentle reminder to review this

**Comment by ali-farooq0** [comment]
@cinlloc can you please sync this PR with `master`, thanks!

**Comment by cinlloc** [comment]
@hanzei sync with `master` done

**Comment by hanzei** [comment]
This is an awesome contribution. I'm looking forward using this myself a lot!

**Pull request merged**